crates/uv-auth/src/middleware.rs

```diff
@@ -0,0 +1,185 @@
+use netrc::Netrc;
+use reqwest::{header::HeaderValue, Request, Response};
+use reqwest_middleware::{Middleware, Next};
+use std::path::Path;
+use task_local_extensions::Extensions;
+use tracing::{debug, warn};
+
+use crate::{
+    keyring::{get_keyring_subprocess_auth, KeyringProvider},
+    store::{AuthenticationStore, Credential},
+};
+
+/// A middleware that adds basic authentication to requests based on the netrc file and the keyring.
+///
+/// Netrc support Based on: <https://github.com/gribouille/netrc>.
+pub struct AuthMiddleware {
+    nrc: Option<Netrc>,
+    keyring_provider: KeyringProvider,
+}
+
+impl AuthMiddleware {
+    pub fn new(keyring_provider: KeyringProvider) -> Self {
+        Self {
+            nrc: Netrc::new().ok(),
+            keyring_provider,
+        }
+    }
+
+    pub fn from_netrc_file(file: &Path, keyring_provider: KeyringProvider) -> Self {
+        Self {
+            nrc: Netrc::from_file(file).ok(),
+            keyring_provider,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Middleware for AuthMiddleware {
+    async fn handle(
+        &self,
+        mut req: Request,
+        _extensions: &mut Extensions,
+        next: Next<'_>,
+    ) -> reqwest_middleware::Result<Response> {
+        let url = req.url().clone();
+        // If the request already has an authorization header, we don't need to do anything.
+        // This gives in-URL credentials precedence over the netrc file.
+        if req.headers().contains_key(reqwest::header::AUTHORIZATION) {
+            if !url.username().is_empty() {
+                AuthenticationStore::save_from_url(&url);
+            }
+            return next.run(req, _extensions).await;
+        }
+
+        // Try auth strategies in order of precedence:
+        if let Some(stored_auth) = AuthenticationStore::get(&url) {
+            // If we've already seen this URL, we can use the stored credentials
```

Comment:
For my own clarification... this will propagate the credentials based purely on `host` etc., right? In other words, the URLs do not need to be identical, is that correct? Just the same "realm" and so on?

Is this generally a safe and accepted thing to do? (That is: propagate the credentials here.) It's a little different than what we do on `main` which is more limited: when we make a request, we propagate credentials to the URLs that are found on the page returned by the request (like `with_url_encoded_auth`, I think).

Reply:
Yeah we're using the "net loc" which is the same "realm". This is compliant with the relevant web standard RFC and AFAIK exactly what pip does. It seems totally fine to me but I appreciate the extra scrutiny.

Reply:
Do you know why we _also_ need to call `with_url_encoded_auth` before the middleware then?

Reply:
I didn't follow the full flow after the `url` is copied - I saw it was being used in `File` structs and didn't want to potentially introduce regression by assuming theses `urls` having the auth copied to were going to be used in the same run of `uv`